### DISCUSSION TITLE
Omit unsupported fields or reject Objects containing unsupported fields for given version of Prometheus

### DISCUSSION THREAD

**Discussion by mviswanathsai** [prometheus-operator/prometheus-operator#6792] [General]
### What is missing? Please describe.

I want to clarify that this discussion can also extend to the monitor resources and other resources as well. Since those resources are stable and this is potentially a breaking change, the conversation might be more meaningful in the context of the `ScrapeConfig` CRD. 

Currently, if a user specifies a field in an object (say `ScrapeConfig`) that is unsupported for the given version of Prometheus, we omit  that field from the generated config. The problem with this being, we usually have no idea that this happened. The user expects something and the generated config does not match it, and the user has no real way to know this except manually checking the generated config.

The alternative to this would be rejecting the object containing the unsupported field with a meaningful error message stating the reason why the object was rejected. 

Pros:
1. Clear: 
Scenario: The user does not see the targets they wanted to, they open the logs and find out that their object was rejected and the reason why it was rejected. 
2. Restrictive: 
This might not necessarily be a pro, but rejecting objects with "illegal" fields rather than silently dealing with it seems a better way to deal with things. 

Cons:
1. As pointed out by @simonpasquier, the newer fields might be only used by a handful of "advanced" users who already know what they are doing and so it might not be necessary to reject the object.
2. These newer fields do not usually make a big difference in the service discovery mechanism and so we might not need to punish a wrongful entry with a rejection. 
3. Might break somethings for existing users.
The rejecting route would be a "breaking" change but in reality if the unsupported field that they specified was omitted in the generated config, then the user does not anyway observe the expected behavior that they hoped for by specifying that field. 

My opinion:
I do not have a strong opinion, but I might be leaning more towards rejecting the field since it is more verbose IMO. We could do a combination of both and do something like
1. Omit the field.
2. Warn the user.
This might be sufficient, but as I already said, most of these fields would probably not cause targets to not show up (there are exceptions however, `Filter` for example) which means the users do not immediately notice the log. 

There was some confusion around the topic and we need to make a decision ASAP, because we will be refactoring the remaining SDs and if we choose to go the way of rejection, we would need to start from the top and refactor again. This is a pain because we would also need to update test cases along with it, on top of moving logic etc. So, the sooner the better.      

CC: @slashpai @ArthurSens @nicolastakashi @AshwinSriram11 @haanhvu 

## Environment
Kubernetes Version:
Prometheus-Operator Version:

**Comment by ArthurSens** [comment]
Thanks for raising your concerns!

Do I understand correctly that we don't even emit a log? What do you think about adding warnings in the log as a first step? I'm afraid that completely rejecting a resource would be a breaking change 😬

**Reply by simonpasquier** [reply]
The operator emits warning logs when it detects a mismatch between the Prometheus version and the setting. The question originated from reviewing changes to the ScrapeConfig CRD for which we could break the contract since it's alpha.

**Comment by AshwinSriram11** [comment]
I am also leaning toward having warning logs rather than rejecting the whole resource. If a log indicates that a field is missing, then the user can figure out the behavior.

>The operator emits warning logs when it detects a mismatch between the Prometheus version and the setting. 

Just want to be clear, currently, there are no logs that directly indicate that a particular field is unsupported.

**Reply by simonpasquier** [reply]
> currently, there are no logs that directly indicate that a particular field is unsupported.

There should be
